### ISSUE TITLE
There is no easy way to differentiate between elements

### ISSUE THREAD

**Issue by Dassderdie** [hpi-sam/digital-fuesim-manv#530]
### Problem

Currently, there is no "nice" way to differentiate between e.g. a `Patient` and a `Material` object.
At the moment, we use two different workarounds for this:

#### Type guards (that could break easily) 
`shared\src\store\action-reducers\utils\calculate-treatments.ts`
```ts
function isPatient(
    element: Mutable<Material | Patient | Personnel>
): element is Mutable<Patient> {
    return (element as Patient).personalInformation !== undefined;
}
```

#### Pass a type with the element

In addition to the element itself, the respective `type` is manually propagated to other functions (much more verbose).

1. 
* "Sender": `frontend\src\app\pages\exercises\exercise\shared\exercise-map\feature-managers\element-manager.ts` - `getElementFromFeature()`
* Receiver: `frontend\src\app\pages\exercises\exercise\shared\exercise-map\feature-managers\transfer-point-feature-manager.ts` - `onFeatureDrop()`

2. 
`frontend\src\app\pages\exercises\exercise\shared\core\drag-element.service.ts` - `TransferTemplate`

##### [`instanceof`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof?retiredLocale=de)
* Sadly, this doesn't work in our case, as objects saved in the state do not have a `prototype`. Remember that our state is an immutable JSON object that can be sent over the network.

##### Add a `type` property to the models
* Each model gets a readonly `type` property (similar to the `type` property in an action like in `shared\src\store\action-reducers\material.ts` - `MoveMaterialAction`)
* This type should be plural (`personnel`, `patients`, ...) to be reusable as the key in the respective Maps in the `ExerciseState` (`shared\src\state.ts`).
* TypeScript has a [discriminated-union feature](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions), this enables easy differentiation between the objects later.

1. Add the `type` property to the following models (`shared\src\models`):
* [ ] Viewport
* [ ] Vehicle
* [ ] Personnel
* [ ] Material
* [ ] MapImage
* [ ] TransferPoint
* [ ] Hospital
* [ ] HospitalPatient
* [ ] AlarmGroup
* [ ] Client
* [ ] PatientCategory
* [ ] VehicleTemplate
* [ ] MapImageTemplate
* [ ] PersonnelTemplate

**Comment by ClFeSc** [comment]
We could use `instanceof` if we used `class-transformer` to transform the objects before and after sending. We already use it for validation, so most things would already be correctly annotated. However, there might arise other problems when we always use this.

**Comment by Dassderdie** [comment]
> We could use `instanceof` if we used `class-transformer` to transform the objects before and after sending. We already use it for validation, so most things would already be correctly annotated. However, there might arise other problems when we always use this.

In general `instanceof` also has other "problems":
* You only get limited autocompletion for the class in comparison to a discriminated union
* Inheritance is handled differently:
````ts
class A { }
class B extends A { }
const a = new A();
const b = new B();
console.log(a instanceof A); // true
console.log(b instanceof A); // true
````


## TLDR:

I believe the discriminated union via a `type` property is a better approach than the current ones.
Therefore this would be an (incremental) improvement.
Using `instanceof` would require a lot of (complicated) refactoring and more discussion. I created #531 for this.

**Comment by ClFeSc** [comment]
I agree with this assessment. At least for now `type` should be better and would be much less work.

**Comment by Dassderdie** [comment]
After this is done, the typings of the droppedElement could also by strictly typed by a union. -> This would probably be a new issue/branch.

**Comment by lukasrad02** [comment]
@Dassderdie you're currently assigned on this. Are you actively working on this issue or would you mind if we would work on this?

**Comment by Dassderdie** [comment]
> @Dassderdie you're currently assigned on this. Are you actively working on this issue or would you mind if we would work on this?

Feel free to work on this. I was still assigned and the title has a TODO as I wasn't sure about whether every element should get a type or YAGNI...
